cfn-resources/serverless-instance/cmd/resource/resource.go

```diff
@@ -0,0 +1,395 @@
+	if currentModel.ProviderSettings == nil {
+		return &mongodbatlas.ServerlessProviderSettings{
+			ProviderName:        constants.Serverless,
+			BackingProviderName: constants.AWS,
+		}
+	}
+	serverlessProviderSettings = &mongodbatlas.ServerlessProviderSettings{
+		BackingProviderName: constants.AWS,
+	}
+	if currentModel.ProviderSettings.ProviderName != nil {
+		serverlessProviderSettings.ProviderName = *currentModel.ProviderSettings.ProviderName
+	}
+	if currentModel.ProviderSettings.RegionName != nil {
+		serverlessProviderSettings.RegionName = *currentModel.ProviderSettings.RegionName
+	}
+	return serverlessProviderSettings
+}
+
+func readServerlessInstance(cluster *mongodbatlas.Cluster) (serverless *Model) {
+	serverless = &Model{}
+	serverless.Name = &cluster.Name
+	serverless.Id = &cluster.ID
+	serverless.ProjectID = &cluster.GroupID
+	if cluster.ProviderSettings != nil {
+		serverless.ProviderSettings = &ServerlessInstanceProviderSettings{
+			ProviderName: &cluster.ProviderSettings.ProviderName,
+			RegionName:   &cluster.ProviderSettings.RegionName,
+		}
+	}
+
+	if cluster.ServerlessBackupOptions != nil {
+		serverless.ContinuousBackupEnabled = cluster.ServerlessBackupOptions.ServerlessContinuousBackupEnabled
+	}
+
+	if cluster.ConnectionStrings != nil {
+		serverless.ConnectionStrings = &ServerlessInstanceConnectionStrings{
+			StandardSrv:     &cluster.ConnectionStrings.StandardSrv,
+			PrivateEndpoint: readPrivateEndpoint(cluster.ConnectionStrings.PrivateEndpoint),
+		}
+	}
+	serverless.CreateDate = &cluster.CreateDate
+	serverless.MongoDBVersion = &cluster.MongoDBVersion
+	serverless.Links = readLinks(cluster.Links)
+	serverless.TerminationProtectionEnabled = cluster.TerminationProtectionEnabled
+	serverless.StateName = &cluster.StateName
+	return serverless
+}
+
+func readPrivateEndpoint(privateEPs []mongodbatlas.PrivateEndpoint) (endPoints []ServerlessInstancePrivateEndpoint) {
+	for i := range privateEPs {
+		var pep = ServerlessInstancePrivateEndpoint{}
+		pep.Endpoints = readPrivateEndpointEndpoints(privateEPs[i].Endpoints)
+		pep.Type = &privateEPs[i].Type
+		pep.SrvConnectionString = &privateEPs[i].SRVConnectionString
+		endPoints = append(endPoints, pep)
+	}
+	return
+}
+
+func readPrivateEndpointEndpoints(peEndpoints []mongodbatlas.Endpoint) (epEndpoints []ServerlessInstancePrivateEndpointEndpoint) {
+	for i := range peEndpoints {
+		epEndpoints = append(epEndpoints, ServerlessInstancePrivateEndpointEndpoint{
+			EndpointId:   &peEndpoints[i].EndpointID,
+			ProviderName: &peEndpoints[i].ProviderName,
+			Region:       &peEndpoints[i].Region,
+		})
+	}
+	return
+}
+
+func readLinks(clsLinks []*mongodbatlas.Link) (links []Link) {
+	for i := range clsLinks {
+		links = append(links, Link{
+			Href: &clsLinks[i].Href,
+			Rel:  &clsLinks[i].Rel,
+		})
+	}
+	return
+}
+
+func serverlessCallback(client *mongodbatlas.Client, currentModel *Model, targtStatus string) (progressEvent handler.ProgressEvent, err error) {
```

Comment:
you dont have to fix this, is just a personal opinion, mixing the delete and the create asynchronous check in a single function, makes in complicated to read, is like both flows are mixed in one function, mybe for future resource we can keep them separated.